### PR TITLE
Render publish plugins abstraction

### PR DESCRIPTION
**Pull request by antirotor**
## Feature

This is attempt to abstract how renders are collected, submitted and published on farm. This is useful for uniting most of the code between hosts.

Idea is to inherit from those abstract plugins in host specific one, implement abstract method and/or override those that make sense for current host.

To unify data collected I used data class concept. There is support in python for data classes from 3.7 but we still need to support legacy python 2 hosts, so I opted for using [attrs](https://www.attrs.org/) module we already have in dependecies as it is pulled by something.

### CODE REVIEW

File: pype/lib/abstract_collect_render.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    @abstractmethod
+    def get_instances(self):
+        """Get all renderable instances and their data.
+
+        Returns:
+            list of :class:`RenderInstance`: All collected renderable instances
+                (like render layers, write nodes, etc.)
+
+        """
+        pass
```
Comment (tokejepsen) [inline pype/lib/abstract_collect_render.py:207]: Maybe raise an `NotImplementedError` here?

Comment (antirotor) [reply pype/lib/abstract_collect_render.py:207]: > Maybe raise an `NotImplementedError` here?

no need for that - `@abstractmethod` will take care of it. You cannot instantiate classes with unimplemented abstract method, that will rise `TypeError: Can't instantiate abstract class Foo with abstract methods bar`

File: pype/lib/abstract_collect_render.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+                frame_start = frame_start_render
+                frame_end = frame_end_render
+                frame_start_handle = frame_start_render
+                frame_end_handle = frame_end_render
+
+            data = {
+                "subset": render_instance.subset,
+                "attachTo": render_instance.attachTo,
+                "setMembers": render_instance.setMembers,
+                "multipartExr": exp_files.multipart,
```
Comment (kalisp) [inline pype/lib/abstract_collect_render.py:166]: Will not work if `exp_files` are list only

File: pype/lib/abstract_collect_render.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    def process(self, context):
+        """Entry point to collector."""
+        for instance in context:
+            # make sure workfile instance publishing is enabled
+            if "workfile" in instance.data["families"]:
+                instance.data["publish"] = True
+
+        self._file_path = context.data["currentFile"].replace("\\", "/")
+
+        render_instances = self.get_instances()
```
Comment (kalisp) [inline pype/lib/abstract_collect_render.py:128]: `context` argument is missing

**Pull request merged**